File: Colors/Colors/ViewController.m
//
//  ViewController.m
//  Colors
//
//  Created by Lukasz Kuczborski on 21.01.2015.
//  Copyright (c) 2015 ACME. All rights reserved.
//

#import "ViewController.h"

static NSString *const RED_UUID   = @"430F6ED4-7E4F-4F63-85C7-845763861A55";
static NSString *const BLUE_UUID  = @"B9407F30-F5F8-466E-AFF9-25556B57FE6D";
static NSString *const GREEN_UUID = @"3A7E7514-A9AC-41F3-A236-3CFCE75BAC95";

@interface ViewController () <BeaconEventHandler>

@property (nonatomic, strong) BeaconHandler *handler;

@end

@implementation ViewController

- (void)viewDidAppear:(BOOL)animated
{
	[super viewDidAppear:animated];
	
	[self setupBeaconHandler];
}

- (IBAction)switchTeam:(UIControl *)sender {
}
- (CGFloat)getComponentValueFromBeacon:(Beacon *)beacon {

	// http://stn.spotfire.com/spotfire_client_help/norm/norm_scale_between_0_and_1.htm
	CGFloat normalized = fabsl(beacon.accuracy - beacon.accuracyMin) / fabsl(beacon.accuracyMax - beacon.accuracyMin);
	
	// some more hints using proximity
	if (beacon.proximity == CLProximityImmediate) {
		normalized += 0.5;
	}
	else if (beacon.proximity == CLProximityUnknown) {
		normalized -= 0.1;
	}
	else if (beacon.proximity == CLProximityNear){
		normalized += 0.1;
	}
	
	// checking the bounds again...
	if (normalized < 0) {
		normalized = 0.0001;
	}
	
	if (normalized > 1) {
		normalized = 1;
	}
	
    return normalized;
}

- (void)animateToColor:(UIColor*)color {

    //TODO: find a nice way to animate this...
	[UIView animateWithDuration:0.5
						  delay:0
						options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionCurveEaseOut
					 animations:^{
						 self.view.layer.backgroundColor = color.CGColor;
					 } completion:nil];
}

#pragma mark - BeaconHandler

- (void)setupBeaconHandler {
	NSArray *regions = @[
						 [[BeaconRegion alloc] initWithUUID:RED_UUID
													   name:@"Estimote Mint Cocktail"],
						 [[BeaconRegion alloc] initWithUUID:BLUE_UUID
													   name:@"Estimote Icy Marshmallow"],
						 [[BeaconRegion alloc] initWithUUID:GREEN_UUID
													   name:@"Estimote Blueberry Pie"]
						 ];
	
	self.handler = [[BeaconHandler alloc] initWithBeaconRegions:regions
													   delegate:self];
	
	[self.handler start];
}

#pragma mark - BeaconEventHandler

- (void)beaconHandler:(BeaconHandler *)handler
	 didUpdateBeacons:(NSArray *)beacons {
	
	UIColor *currentColor = [UIColor colorWithCGColor:self.view.layer.backgroundColor];
	CGFloat r = 0.f;
	CGFloat g = 0.f;
	CGFloat b = 0.f;
	CGFloat a = 1.f;
	[currentColor getRed:&r green:&g blue:&b alpha:&a];
	r *= 0.9; g *= 0.9; b *= 0.9; 
	for (Beacon *beacon in beacons) {
		if ([beacon.uuid isEqualToString:RED_UUID]) {
			r = [self getComponentValueFromBeacon:beacon];
		} else if ([beacon.uuid isEqualToString:GREEN_UUID]) {
			g = [self getComponentValueFromBeacon:beacon];
		} else if ([beacon.uuid isEqualToString:BLUE_UUID]) {
			b = [self getComponentValueFromBeacon:beacon];
		}
	}
	
	NSLog(@"%s\t :  R:%.2f G:%.2f B:%.2f", __PRETTY_FUNCTION__, r, g, b);
	[self animateToColor:[UIColor colorWithRed:r green:g blue:b alpha:a]];
}

@end

File: Colors_Swift/Colors_Swift/ViewController.swift
//
//  ViewController.swift
//  Colors_Swift
//
//  Created by Lukasz Kuczborski on 28.07.2016.
//  Copyright © 2016 Beacon Tech. All rights reserved.
//

import UIKit
import PorkKit

let RED_UUID = "430F6ED4-7E4F-4F63-85C7-845763861A55"
let BLUE_UUID = "B9407F30-F5F8-466E-AFF9-25556B57FE6D"
let GREEN_UUID = "3A7E7514-A9AC-41F3-A236-3CFCE75BAC95"

class ViewController: UIViewController {

    var handler: PorkHandler?

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        self.setupPorkHandler()
    }

    // MARK: PorkHandler

    func setupPorkHandler() {

        guard let redRegion: PorkRegion = try? PorkRegion(name: "Estimote Mint Cocktail", uuid: RED_UUID),
              let blueRegion: PorkRegion = try? PorkRegion(name: "Estimote Icy Marshmallow", uuid: BLUE_UUID),
              let greenRegion: PorkRegion = try? PorkRegion(name: "Estimote Blueberry Pie", uuid: GREEN_UUID)
        else {
            return
        }

        let regions = [redRegion, blueRegion, greenRegion]

        do {
            handler = try PorkHandler(porkRegions: regions, porkEventHandler: self)
            try handler?.start()
        } catch {
            print("Error: \(error)")
        }
    }

    func getComponentValueFromBeacon(_ beacon: Pork) -> CGFloat {

        var normalized: CGFloat = 0.5

        // some more hints using proximity
        if beacon.proximity == .immediate {
            normalized += 0.5
        }
        else if beacon.proximity == .unknown {
            normalized -= 0.1
        }
        else if beacon.proximity == .near {
            normalized += 0.1
        }

        // checking the bounds again...
        if (normalized < 0) {
            normalized = 0.0001
        }
        
        if (normalized > 1) {
            normalized = 1
        }
        
        return normalized
    }

    func animateToColor(_ color: UIColor) {
        // TODO: find a nice way to animate this...
        UIView.animate(withDuration: 0.5,
                       delay: 0.0,
                       options: [.beginFromCurrentState, .curveEaseOut],
                       animations: {
                        self.view.layer.backgroundColor = color.cgColor
                        },
                       completion: nil)
    }
}

// MARK: PorkEventHandler

extension ViewController: PorkEventHandler {

    func porkHandlerDidUpdateBeacons(_ porkHandler: PorkHandler, beacons: [Pork]) {

        let currentColor: UIColor = UIColor(cgColor: view.layer.backgroundColor!)
        var r: CGFloat = 0.0
        var g: CGFloat = 0.0
        var b: CGFloat = 0.0
        var a: CGFloat = 1.0

        currentColor.getRed(&r, green: &g, blue: &b, alpha: &a)
        for beacon in beacons {
            switch beacon.uuid {
            case RED_UUID:
                r = getComponentValueFromBeacon(beacon)
            case GREEN_UUID:
                g = getComponentValueFromBeacon(beacon)
            case BLUE_UUID:
                b = getComponentValueFromBeacon(beacon)
            default:
                break
            }
        }

        print("R: \(r), G: \(g), B: \(b)")
        self.animateToColor(UIColor(red: r, green: g, blue: b, alpha: a))
    }
}
